ODIN_II/SRC/ast_elaborate.cpp

```diff
@@ -847,17 +847,23 @@ void convert_2D_to_1D_array(ast_node_t **var_declare, STRING_CACHE_LIST *local_s
 	ast_node_t *new_node = create_tree_node_number(addr_chunk_size, (*var_declare)->children[0]->line_number, (*var_declare)->children[0]->file_number);
 
 	STRING_CACHE *local_param_table_sc = local_string_cache_list->local_param_table_sc;
-	long sc_spot;
 
 	char *temp_string = make_chunk_size_name(instance_name_prefix, name);
 
-	if ((sc_spot = sc_add_string(local_param_table_sc, temp_string)) == -1)
+	long sc_spot = sc_add_string(local_param_table_sc, temp_string);
+	if (sc_spot == -1)
```

Comment:
@ademmings I don't think this check will work? it should error if the data is not NULL if I'm not mistaken

Reply:
Yeah that sounds about right... This piece was copy/pasted from something we wrote two months ago for a quick bug fix, and we've since discovered that sc_add_string() doesn't ever return -1, so that error check is wrong. (There were lots of error checks like that in the code that we didn't know were wrong...)

Rather than 
```if (sc_spot == -1)``` 
it should actually be 
```if (local_param_table_sc->data[sc_spot] != NULL)```
and throw the error message if there's already a node in that spot, i.e. if the user defined a variable with the same name. Otherwise, the free you're doing later on would free the user's variable and replace it with the new array name that we are creating internally, rather than throwing the error for the naming conflict and quitting as intended.

Reply:
I'll go ahead and change that then.. that will cause matmul.v to crash though, so should I also disable that test?

Reply:
I just thought about it and yeah that would crash if it's being called by update_string_cache()... We shouldn't have to disable the test, but we should rather have either some kind of check to see if that array has already been expanded and added, or update the reference in the string cache in place, because I changed the code so that everything in the string cache is a copy and not the original, meaning that variables have to be updated twice (hence update_string_cache()).

Reply:
Unless it crashes on something else and not the error message? In which case we have to look into it more

Reply:
Seems like it could be a bug with 2D arrays. Looks like it's creating a second node with the same contents and replacing the old one that's already in the string cache.. unless I'm not understanding properly here?
![Screenshot from 2019-08-01 13-18-14](https://user-images.githubusercontent.com/50643234/62309942-d7165000-b45e-11e9-925e-e84b60605662.png)

![Screenshot from 2019-08-01 12-58-11](https://user-images.githubusercontent.com/50643234/62309639-4b9cbf00-b45e-11e9-8c03-e505e29228a4.png)

Reply:
That's what I was saying about variables being updated twice. There's one copy of the node in the AST, and a second copy in the string cache, and both get resolved individually. If you look at the backtrace you might notice that the function's being called from two different places, too. So that's the issue that we have to get around - it should only add it to the string cache once, but if both copies of the node exist, they both need to be updated. 

JP and I were talking about a proper solution for this, and I think we decided on changing the AST so that the node doesn't ever appear twice - once it's copied into the string cache, it can be removed from the AST, because we don't need it there anymore. If you look at create_symbol_table_for_module() in netlist_create_from_ast(),  you should be able to find where I was making deep copies of the variables for the string cache (the bit with stored_var_declare) - there just needs to be a couple of lines added to remove it from the AST once it's in the string cache. Once there's only one copy of the node, it should only go through convert_2D_to_1D once. I'm hoping that that will fix the problem. :)